### PR TITLE
EntryPoint: avoid costly pyparsing for dummy requirement string

### PR DESCRIPTION
**Pull request by tmmorin**
In the case where an EntryPoint is created with no extra, a Requirement
object is instanciated with a dummy requirement string 'x[]', which incurs
a useless significant cost due to the use of pyparsing to parse this string.

This change skips the instanciation of a Requirements objects in the
case where extra is empty.

See github issue #1132 for more details, including information on the
performance improvement brought by this change.

### CODE REVIEW

File: pkg_resources/__init__.py [hunk 1]
```diff
@@ -2286,7 +2286,10 @@ def __init__(self, name, module_name, attrs=(), extras=(), dist=None):
         self.name = name
         self.module_name = module_name
         self.attrs = tuple(attrs)
-        self.extras = Requirement.parse(("x[%s]" % ','.join(extras))).extras
+        if extras:
+            self.extras = Requirement.parse(("x[%s]" % ','.join(extras))).extras
```
Comment (webknjaz) [inline pkg_resources/__init__.py]: It looks like `Requirement.parse` and underlying `parse_requirements` aren't so expensive. But anyway I'd add this check there.

**Comment by jaraco** [comment]
Thanks very much for the ticket with a detailed description of the problem and a clear indication of a performance impact. This PR is an excellent example showing how to communicate a problem and a proposed solution.

I do have some critiques that I'd like to see if we can address.

My main concern lies with the lack of comment or test protecting this new behavior. I could easily see myself optimizing this change away, not realizing the affect on performance. If it would be possible to write a test, that would be best... but a separate function documenting the intention would also be good.

My preference, though, would be for packaging to expose a 'parse_extras' function or method that could be readily invoked without the expense of parsing a dummy requirement, such that the setuptools implementation was simple and direct. Let's explore that first.

**Comment by tmmorin** [comment]
Thanks @jaraco and I agree that documenting this properly is required to avoid regressions.
I also agree that having this done right in `packaging` would be even better.

I'm I interpreting your suggestion correctly as meaning to add a `parse_extras` method to `packaging.requirements.Requirement` [1].

But then I fail to see how to write such a function: in the case where extras is non empty, it would have to call the "regular" parse(<generated requirement string for the extras>).   But `packaging.requirements.Requirement` does *not* have a `parse` method, the `parse` method used at [2] is in fact added in the wrapper at [3].

Given that, I would prefer adding a `parse_extras` method in this wrapper, with proper documentation on why we shortcut the case where `extras` is empty.

[1] https://github.com/pypa/packaging/blob/master/packaging/requirements.py#L78
[2] https://github.com/tmmorin/setuptools/blob/94fa402a207e27139499d3347a69a7d8f7036939/pkg_resources/__init__.py#L2290
[3] https://github.com/tmmorin/setuptools/blob/94fa402a207e27139499d3347a69a7d8f7036939/pkg_resources/__init__.py#L2950-L2953

**Comment by jaraco** [comment]
In d1f16fdd, I added one small optimization, bypassing much of the unnecessary processing in `pkg_resources.Requirement.__init__`, though I suspect from what I see in the original ticket that profiling revealed `pyparsing` as the primary culprit, so this structural optimization won't provide the desired performance optimization.

**Comment by jaraco** [comment]
In 4ee5c651, I found what I think is an even better solution. If there was a purpose behind the serialization/parsing step, it's not captured by comments nor tests, so it's subject to change.

**Pull request closed without merging**